File: ADE-stable-1.0/src/core/agents/agent_registry.py
from typing import Dict, Any, Optional, List
import logging
from datetime import datetime

from ..agent.base import BaseAgent, AgentCapability
from ...agents.development.workflow_manager import WorkflowManagerAgent
from ...agents.development.planner import PlannerAgent
from ...agents.leadership.architect import ArchitectAgent

logger = logging.getLogger(__name__)

class AgentRegistry:
    """Registry for managing ADE platform agents"""
    
    def __init__(self):
        self.agents: Dict[str, BaseAgent] = {}
        self.agent_capabilities: Dict[str, List[AgentCapability]] = {}
        self.active_collaborations: Dict[str, Dict[str, Any]] = {}
        self.logger = logging.getLogger(__name__)
        
    def register_agent(self, agent: BaseAgent) -> bool:
        """Register a new agent in the registry"""
        try:
            if agent.agent_id in self.agents:
                self.logger.warning(f"Agent {agent.agent_id} is already registered")
                return False
                
            self.agents[agent.agent_id] = agent
            self.agent_capabilities[agent.agent_id] = agent.capabilities
            self.logger.info(f"Successfully registered agent {agent.agent_id} with capabilities {agent.capabilities}")
            return True
        except Exception as e:
            self.logger.error(f"Error registering agent: {str(e)}")
            return False
            
    def unregister_agent(self, agent_id: str) -> bool:
        """Unregister an agent from the registry"""
        try:
            if agent_id not in self.agents:
                self.logger.warning(f"Agent {agent_id} is not registered")
                return False
                
            del self.agents[agent_id]
            del self.agent_capabilities[agent_id]
            self.logger.info(f"Successfully unregistered agent {agent_id}")
            return True
        except Exception as e:
            self.logger.error(f"Error unregistering agent: {str(e)}")
            return False
            
    def get_agent(self, agent_id: str) -> Optional[BaseAgent]:
        """Get an agent by ID"""
        return self.agents.get(agent_id)
        
    def get_agents_by_capability(self, capability: AgentCapability) -> List[BaseAgent]:
        """Get all agents with a specific capability"""
        return [
            agent for agent_id, agent in self.agents.items()
            if capability in self.agent_capabilities[agent_id]
        ]
        
    def initialize_workflow_agents(self, provider_registry: Any) -> None:
        """Initialize workflow-related agents"""
        try:
            # Initialize workflow manager agent
            workflow_manager = WorkflowManagerAgent(
                agent_id="workflow_manager_1",
                provider_registry=provider_registry
            )
            self.register_agent(workflow_manager)
            
            # Initialize planner agent
            planner = PlannerAgent(
                agent_id="planner_1",
                provider_registry=provider_registry
            )
            self.register_agent(planner)
            
            # Initialize architect agent
            architect = ArchitectAgent(
                agent_id="architect_1",
                provider_registry=provider_registry
            )
            self.register_agent(architect)
            
            self.logger.info("Successfully initialized workflow agents")
        except Exception as e:
            self.logger.error(f"Error initializing workflow agents: {str(e)}")
            
    def start_collaboration(self, source_agent_id: str, target_agent_id: str, task: Dict[str, Any]) -> Dict[str, Any]:
        """Start a collaboration between two agents"""
        try:
            if source_agent_id not in self.agents or target_agent_id not in self.agents:
                return {"status": "error", "message": "One or both agents not found"}
                
            source_agent = self.agents[source_agent_id]
            target_agent = self.agents[target_agent_id]
            
            collaboration_id = f"collab_{datetime.now().strftime('%Y%m%d_%H%M%S')}"
            self.active_collaborations[collaboration_id] = {
                "source_agent": source_agent_id,
                "target_agent": target_agent_id,
                "task": task,
                "start_time": datetime.now(),
                "status": "active"
            }
            
            return {
                "status": "success",
                "collaboration_id": collaboration_id,
                "message": "Collaboration started successfully"
            }
        except Exception as e:
            self.logger.error(f"Error starting collaboration: {str(e)}")
            return {"status": "error", "message": str(e)}
            
    def end_collaboration(self, collaboration_id: str) -> Dict[str, Any]:
        """End an active collaboration"""
        try:
            if collaboration_id not in self.active_collaborations:
                return {"status": "error", "message": "Collaboration not found"}
                
            collaboration = self.active_collaborations[collaboration_id]
            collaboration["end_time"] = datetime.now()
            collaboration["status"] = "completed"
            
            return {
                "status": "success",
                "message": "Collaboration ended successfully"
            }
        except Exception as e:
            self.logger.error(f"Error ending collaboration: {str(e)}")
            return {"status": "error", "message": str(e)}
            
    def get_collaboration_status(self, collaboration_id: str) -> Dict[str, Any]:
        """Get the status of a collaboration"""
        try:
            if collaboration_id not in self.active_collaborations:
                return {"status": "error", "message": "Collaboration not found"}
                
            return {
                "status": "success",
                "data": self.active_collaborations[collaboration_id]
            }
        except Exception as e:
            self.logger.error(f"Error getting collaboration status: {str(e)}")
            return {"status": "error", "message": str(e)} 

File: backend/src/services/agent/AgentRegistry.ts
import { singleton } from 'tsyringe';
import { Logger } from '../logger/Logger';
import { AgentService, AgentConfig, AgentCapability } from './AgentService';
import { Subject, Observable } from 'rxjs';
import { WebSocket } from 'ws';

export interface AgentRegistration {
  agentId: string;
  timestamp: Date;
  status: 'active' | 'inactive' | 'error';
  error?: string;
  previewUrl?: string;
  collaborators?: string[];
  lastActivity?: Date;
}

export interface AgentActivity {
  agentId: string;
  action: string;
  timestamp: Date;
  data?: any;
}

export interface CollaborationRequest {
  sourceAgentId: string;
  targetAgentId: string;
  action: 'start' | 'end';
  context?: any;
}

@singleton()
export class AgentRegistry {
  private static instance: AgentRegistry;
  private registrations: Map<string, AgentRegistration>;
  private logger: Logger;
  private agentService: AgentService;
  private activitySubject: Subject<AgentActivity>;
  private collaborationSubject: Subject<CollaborationRequest>;
  private previewSockets: Map<string, Set<WebSocket>>;

  constructor(logger: Logger, agentService: AgentService) {
    this.logger = logger;
    this.agentService = agentService;
    this.registrations = new Map();
    this.activitySubject = new Subject();
    this.collaborationSubject = new Subject();
    this.previewSockets = new Map();
  }

  public static getInstance(logger: Logger, agentService: AgentService): AgentRegistry {
    if (!AgentRegistry.instance) {
      AgentRegistry.instance = new AgentRegistry(logger, agentService);
    }
    return AgentRegistry.instance;
  }

  public async registerAgent(config: AgentConfig): Promise<AgentRegistration> {
    try {
      await this.agentService.registerAgent(config);
      
      const registration: AgentRegistration = {
        agentId: config.id,
        timestamp: new Date(),
        status: 'active',
        collaborators: [],
        lastActivity: new Date()
      };
      
      this.registrations.set(config.id, registration);
      this.logger.info(`Agent ${config.id} registered successfully in registry`);
      
      // Initialize preview socket set for this agent
      this.previewSockets.set(config.id, new Set());
      
      // Emit registration activity
      this.emitActivity({
        agentId: config.id,
        action: 'register',
        timestamp: new Date(),
        data: { config }
      });
      
      return registration;
    } catch (error) {
      const errorMsg = error instanceof Error ? error.message : 'Unknown error';
      const registration: AgentRegistration = {
        agentId: config.id,
        timestamp: new Date(),
        status: 'error',
        error: errorMsg
      };
      
      this.registrations.set(config.id, registration);
      this.logger.error(`Failed to register agent ${config.id}:`, error);
      
      return registration;
    }
  }

  public async unregisterAgent(agentId: string): Promise<void> {
    try {
      await this.agentService.unregisterAgent(agentId);
      this.registrations.delete(agentId);
      
      // Close and clear all preview sockets
      const sockets = this.previewSockets.get(agentId);
      if (sockets) {
        sockets.forEach(socket => socket.close());
        this.previewSockets.delete(agentId);
      }
      
      // Emit unregister activity
      this.emitActivity({
        agentId,
        action: 'unregister',
        timestamp: new Date()
      });
      
      this.logger.info(`Agent ${agentId} unregistered successfully from registry`);
    } catch (error) {
      this.logger.error(`Failed to unregister agent ${agentId}:`, error);
      throw error;
    }
  }

  public getRegistration(agentId: string): AgentRegistration | undefined {
    return this.registrations.get(agentId);
  }

  public getAllRegistrations(): AgentRegistration[] {
    return Array.from(this.registrations.values());
  }

  public async getActiveAgents(): Promise<AgentConfig[]> {
    const activeAgents: AgentConfig[] = [];
    
    for (const [agentId, registration] of this.registrations) {
      if (registration.status === 'active') {
        const agent = this.agentService.getAgent(agentId);
        if (agent && agent.isActive) {
          activeAgents.push(agent);
        }
      }
    }
    
    return activeAgents;
  }

  public async findAgentsByCapability(capability: string): Promise<AgentConfig[]> {
    const agents = await this.getActiveAgents();
    return agents.filter(agent => 
      agent.capabilities.some(cap => cap.supportedActions.includes(capability))
    );
  }

  public async updateAgentStatus(
    agentId: string,
    status: 'active' | 'inactive' | 'error',
    error?: string
  ): Promise<void> {
    const registration = this.registrations.get(agentId);
    if (!registration) {
      throw new Error(`Agent ${agentId} not found in registry`);
    }

    registration.status = status;
    registration.error = error;
    registration.lastActivity = new Date();
    
    this.emitActivity({
      agentId,
      action: 'status_update',
      timestamp: new Date(),
      data: { status, error }
    });
    
    this.logger.info(`Updated status for agent ${agentId} to ${status}`);
  }

  public subscribeToActivity(): Observable<AgentActivity> {
    return this.activitySubject.asObservable();
  }

  public subscribeToCollaboration(): Observable<CollaborationRequest> {
    return this.collaborationSubject.asObservable();
  }

  public async startCollaboration(request: CollaborationRequest): Promise<void> {
    const sourceReg = this.registrations.get(request.sourceAgentId);
    const targetReg = this.registrations.get(request.targetAgentId);

    if (!sourceReg || !targetReg) {
      throw new Error('One or both agents not found');
    }

    if (!sourceReg.collaborators) sourceReg.collaborators = [];
    if (!targetReg.collaborators) targetReg.collaborators = [];

    sourceReg.collaborators.push(request.targetAgentId);
    targetReg.collaborators.push(request.sourceAgentId);

    this.collaborationSubject.next(request);
    
    this.logger.info(`Started collaboration between agents ${request.sourceAgentId} and ${request.targetAgentId}`);
  }

  public async endCollaboration(request: CollaborationRequest): Promise<void> {
    const sourceReg = this.registrations.get(request.sourceAgentId);
    const targetReg = this.registrations.get(request.targetAgentId);

    if (!sourceReg || !targetReg) {
      throw new Error('One or both agents not found');
    }

    if (sourceReg.collaborators) {
      sourceReg.collaborators = sourceReg.collaborators.filter(id => id !== request.targetAgentId);
    }
    if (targetReg.collaborators) {
      targetReg.collaborators = targetReg.collaborators.filter(id => id !== request.sourceAgentId);
    }

    this.collaborationSubject.next(request);
    
    this.logger.info(`Ended collaboration between agents ${request.sourceAgentId} and ${request.targetAgentId}`);
  }

  public registerPreviewSocket(agentId: string, socket: WebSocket): void {
    let sockets = this.previewSockets.get(agentId);
    if (!sockets) {
      sockets = new Set();
      this.previewSockets.set(agentId, sockets);
    }
    sockets.add(socket);

    // Handle socket closure
    socket.on('close', () => {
      this.unregisterPreviewSocket(agentId, socket);
    });
  }

  public unregisterPreviewSocket(agentId: string, socket: WebSocket): void {
    const sockets = this.previewSockets.get(agentId);
    if (sockets) {
      sockets.delete(socket);
      if (sockets.size === 0) {
        this.previewSockets.delete(agentId);
      }
    }
  }

  public broadcastPreviewUpdate(agentId: string, data: any): void {
    const sockets = this.previewSockets.get(agentId);
    if (sockets) {
      const message = JSON.stringify({
        type: 'preview_update',
        agentId,
        data,
        timestamp: new Date()
      });

      sockets.forEach(socket => {
        if (socket.readyState === WebSocket.OPEN) {
          socket.send(message);
        }
      });
    }
  }

  private emitActivity(activity: AgentActivity): void {
    this.activitySubject.next(activity);
  }
} 